### ISSUE TITLE
`assert_smaller` doesn't work with --nobuiltins --nobasepkgs --noprelude

### ISSUE THREAD

**Issue by VictorTaelin** [idris-lang/Idris-dev#4415]
# Steps to Reproduce

Compile the following file with `--nobasepkgs --noprelude --nobuiltins`:

```idris
%default total

data Nat : Type where
  Z : Nat
  S : Nat -> Nat

assert_smaller : (x : a) -> (y : b) -> b
assert_smaller x y = y

fromInteger : Integer -> Nat
fromInteger 0 = Z
fromInteger n = case prim__gtBigInt n 0 of
  0 => Z
  x => S (fromInteger (assert_smaller n (prim__subBigInt n 1)))
```

# Expected Behavior

Should work.

# Observed Behavior

Error:

```
  |
11 | fromInteger 0 = Z
   | ~~~~~~~~~~~~~~~~~
Main.fromInteger is possibly not total due to recursive path Main.fromInteger --> Main.fromInteger
```

**Comment by MaiaVictor** [comment]
Note that Idris's REPL also doesn't work with only builtins, but no prelude / basepkgs. If I run `idris --nobasepkgs --noprelude`, I get the error:

```
Can't find import Builtins
```

Perhaps the right way to do it is to have a local Builtins.idr file and open Idris with `idris --nobasepkgs --noprelude`? If so, some documentation and a better error message than the one above would be desirable.

**Comment by MaiaVictor** [comment]
Another solution suggested by clayrat on Slack is to build my own `Builtins.idr` file and then import it. That file must have `%unqualified` on top. For example:

```
%access public export
%default total

assert_smaller : (x : a) -> (y : b) -> b
assert_smaller x y = y
```

If you import the file above (`import Builtins`) then `assert_smaller` will behave as desired, magically removing compiler complaints of the non-totality of functions over primitives.

**Comment by jfdm** [comment]
@MaiaVictor We sadly don't have much documentation about creating your own prelude, this needs improving. However, it is my opinion what you are seeing is *expected* behaviour. IIRC, the Idris compiler *needs* to know about a set of built-in functions. These are presented in `Builtins.idr` itself. The `--nobuiltins` option is there really for bootstrapping the compilation of `base` and `prelude`, but also for those wishing to roll their own Builtins.

I think an acceptable resolution to your issue would be better documentation on how to roll your own `Builtins`, and own `prelude` and `base`. I sadly do not have much time to look at this right now. However, if you want to contribute some documentation based on your experiences then that would be great.

**Comment by MaiaVictor** [comment]
I disagree. If, with prelude and builtins, it loads the default builtins.idr, why, without Prelude, it wouldn't do the same? In any case, IMO, just replacing the `Can't find import Builtins` message by `Can't find a local import Builtins.idr file` or something similar would help a lot.

**Comment by jfdm** [comment]
@MaiaVictor the experience you might be having maybe be symptomatic of issue #2370.
